src/wallet/rpcwallet.cpp

```diff
@@ -634,21 +752,21 @@ UniValue setlabel(const JSONRPCRequest& request)
         throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Invalid PIVX address");
 
 
-    std::string label;
-    if (request.params.size() > 1)
-        label = LabelFromValue(request.params[1]);
+    std::string label = LabelFromValue(request.params[1]);
 
-    // Only add the label if the address is yours.
     if (IsMine(*pwalletMain, dest)) {
-        // Detect when changing the label of an address that is the 'unused current key' of another label:
+        // Detect when changing the label of an address that is the receiving address of another label:
+        // If so, delete the account record for it. Labels, unlike addresses, can be deleted,
+        // and if we wouldn't do this, the record would stick around forever.
         if (pwalletMain->mapAddressBook.count(dest)) {
             std::string old_label = pwalletMain->mapAddressBook[dest].name;
-            if (dest == GetLabelDestination(old_label))
-                GetLabelDestination(old_label, true);
+            if (old_label != label && dest == GetLabelDestination(old_label))
+                pwalletMain->DeleteLabel(old_label);
         }
         pwalletMain->SetAddressBook(dest, label, AddressBook::AddressBookPurpose::RECEIVE);
```

Comment:
(commit 2392b7598899ac233be0d5c6ab1019acbb03eb0f) This outright sets the purpose to `RECEIVE`, even for an address that might have been saved for a different purpose (i.e. `setlabel` would change the label *and* the purpose).

Reply:
updated the RPC command to pass an empty `purpose` string, resulting in a no purpose change.